File: app/src/test/java/org/alkaline/taskbrain/data/TabStateTest.kt
package org.alkaline.taskbrain.data

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

class TabStateTest {

    private fun tab(noteId: String, displayText: String = noteId) =
        RecentTab(noteId = noteId, displayText = displayText)

    private fun current(noteId: String, displayText: String = noteId) =
        TabState.CurrentTab(noteId = noteId, displayText = displayText)

    // --- computeDisplayTabs ---

    @Test
    fun `computeDisplayTabs pins currentTab in slot 0 and dedupes the shared list`() {
        val result = TabState.computeDisplayTabs(
            current("c"),
            listOf(tab("a"), tab("b"), tab("c"), tab("d")),
        )
        assertEquals(listOf("c", "a", "b", "d"), result.map { it.noteId })
    }

    @Test
    fun `computeDisplayTabs caps at MAX_DISPLAYED total`() {
        val shared = listOf(tab("a"), tab("b"), tab("c"), tab("d"), tab("e"), tab("f"))
        val result = TabState.computeDisplayTabs(current("z"), shared)
        assertEquals(MAX_DISPLAYED, result.size)
        assertEquals("z", result.first().noteId)
        assertEquals(listOf("a", "b", "c", "d"), result.drop(1).map { it.noteId })
    }

    @Test
    fun `computeDisplayTabs keeps current pinned even when not in shared`() {
        // Other devices' writes can push this device's current past the buffer.
        val result = TabState.computeDisplayTabs(
            current("orphan"),
            listOf(tab("a"), tab("b"), tab("c"), tab("d")),
        )
        assertEquals(listOf("orphan", "a", "b", "c", "d"), result.map { it.noteId })
    }

    @Test
    fun `computeDisplayTabs with null current returns shared as-is`() {
        val shared = listOf(tab("a"), tab("b"), tab("c"), tab("d"), tab("e"), tab("f"))
        val result = TabState.computeDisplayTabs(null, shared)
        assertEquals(listOf("a", "b", "c", "d", "e"), result.map { it.noteId })
    }

    @Test
    fun `computeDisplayTabs uses currentTab displayText when shared has stale text`() {
        val result = TabState.computeDisplayTabs(
            current("c", "Fresh title"),
            listOf(tab("c", "Stale title"), tab("a")),
        )
        assertEquals("Fresh title", result.first().displayText)
    }

    // --- nextNoteIdAfterRemove ---

    @Test
    fun `nextNoteIdAfterRemove returns next-most-recent`() {
        assertEquals(
            "b",
            TabState.nextNoteIdAfterRemove("a", listOf(tab("a"), tab("b"), tab("c"))),
        )
    }

    @Test
    fun `nextNoteIdAfterRemove skips the removed id even when not at front`() {
        assertEquals(
            "a",
            TabState.nextNoteIdAfterRemove("b", listOf(tab("a"), tab("b"), tab("c"))),
        )
    }

    @Test
    fun `nextNoteIdAfterRemove returns null when shared has nothing else`() {
        assertNull(TabState.nextNoteIdAfterRemove("a", listOf(tab("a"))))
    }

    @Test
    fun `nextNoteIdAfterRemove returns null when shared is empty`() {
        assertNull(TabState.nextNoteIdAfterRemove("a", emptyList()))
    }

    // --- closeTabNavigationTarget ---

    @Test
    fun `closeTab navigates to next-most-recent when closing current`() {
        val tabs = listOf(tab("a"), tab("b"), tab("c"))
        val result = TabState.closeTabNavigationTarget(tabs, "b", "b")
        assertEquals(CloseTabResult.SwitchTo("a"), result)
    }

    @Test
    fun `closeTab navigates back when closing only tab`() {
        val result = TabState.closeTabNavigationTarget(listOf(tab("a")), "a", "a")
        assertTrue(result is CloseTabResult.NavigateBack)
    }

    @Test
    fun `closeTab stays on current when closing non-current tab`() {
        val tabs = listOf(tab("a"), tab("b"))
        val result = TabState.closeTabNavigationTarget(tabs, "b", "a")
        assertTrue(result is CloseTabResult.StayOnCurrent)
    }

    // --- extractDisplayText ---

    @Test
    fun `extractDisplayText uses first line`() {
        assertEquals("Hello", TabState.extractDisplayText("Hello\nWorld"))
    }

    @Test
    fun `extractDisplayText removes alarm symbol`() {
        assertEquals("Task", TabState.extractDisplayText("⏰ Task"))
    }

    @Test
    fun `extractDisplayText trims whitespace`() {
        assertEquals("Hello", TabState.extractDisplayText("  Hello  "))
    }

    @Test
    fun `extractDisplayText truncates long text`() {
        val long = "A".repeat(20)
        val result = TabState.extractDisplayText(long)
        assertEquals(12, result.length)
        assert(result.endsWith("…"))
    }

    @Test
    fun `extractDisplayText returns New Note for empty content`() {
        assertEquals("New Note", TabState.extractDisplayText(""))
    }

    @Test
    fun `extractDisplayText returns New Note for whitespace-only content`() {
        assertEquals("New Note", TabState.extractDisplayText("   "))
    }

    @Test
    fun `extractDisplayText returns New Note for alarm-symbol-only content`() {
        assertEquals("New Note", TabState.extractDisplayText("⏰"))
    }
}

File: web/src/__tests__/data/TabState.test.ts
import { describe, it, expect } from 'vitest'
import {
  computeDisplayTabs,
  nextNoteIdAfterRemove,
  extractDisplayText,
  MAX_DISPLAYED,
} from '../../data/TabState'
import type { RecentTab } from '../../data/RecentTabsRepository'

function tab(noteId: string, displayText = noteId): RecentTab {
  return { noteId, displayText, lastAccessedAt: null }
}

describe('computeDisplayTabs', () => {
  it('pins currentTab in slot 0 and dedupes the shared list', () => {
    const result = computeDisplayTabs(tab('c'), [tab('a'), tab('b'), tab('c'), tab('d')])
    expect(result.map((t) => t.noteId)).toEqual(['c', 'a', 'b', 'd'])
  })

  it('caps the displayed list at MAX_DISPLAYED total', () => {
    const shared = [tab('a'), tab('b'), tab('c'), tab('d'), tab('e'), tab('f')]
    const result = computeDisplayTabs(tab('z'), shared)
    expect(result).toHaveLength(MAX_DISPLAYED)
    expect(result[0]!.noteId).toBe('z')
    expect(result.slice(1).map((t) => t.noteId)).toEqual(['a', 'b', 'c', 'd'])
  })

  it('keeps current pinned even when not in the shared list', () => {
    // Other devices' writes can push this device's current past the buffer.
    const shared = [tab('a'), tab('b'), tab('c'), tab('d')]
    const result = computeDisplayTabs(tab('orphan'), shared)
    expect(result.map((t) => t.noteId)).toEqual(['orphan', 'a', 'b', 'c', 'd'])
  })

  it('returns the shared list as-is when no current tab', () => {
    const shared = [tab('a'), tab('b'), tab('c'), tab('d'), tab('e'), tab('f')]
    const result = computeDisplayTabs(null, shared)
    expect(result.map((t) => t.noteId)).toEqual(['a', 'b', 'c', 'd', 'e'])
  })

  it('uses currentTab displayText (not the shared entry) so an in-flight title edit shows immediately', () => {
    const result = computeDisplayTabs(
      tab('c', 'Fresh title'),
      [tab('c', 'Stale title'), tab('a')],
    )
    expect(result[0]!.displayText).toBe('Fresh title')
  })
})

describe('nextNoteIdAfterRemove', () => {
  it('returns the next-most-recent shared entry', () => {
    expect(nextNoteIdAfterRemove('a', [tab('a'), tab('b'), tab('c')])).toBe('b')
  })

  it('skips the removed id even if not at the front', () => {
    expect(nextNoteIdAfterRemove('b', [tab('a'), tab('b'), tab('c')])).toBe('a')
  })

  it('returns null when shared has nothing else', () => {
    expect(nextNoteIdAfterRemove('a', [tab('a')])).toBeNull()
  })

  it('returns null when shared is empty', () => {
    expect(nextNoteIdAfterRemove('a', [])).toBeNull()
  })
})

describe('extractDisplayText', () => {
  it('uses first line', () => {
    expect(extractDisplayText('Hello\nWorld')).toBe('Hello')
  })

  it('removes alarm symbol', () => {
    expect(extractDisplayText('⏰ Task')).toBe('Task')
  })

  it('trims whitespace', () => {
    expect(extractDisplayText('  Hello  ')).toBe('Hello')
  })

  it('truncates long text', () => {
    const long = 'A'.repeat(20)
    const result = extractDisplayText(long)
    expect(result).toHaveLength(12)
    expect(result.endsWith('…')).toBe(true)
  })

  it('returns (empty) for empty content', () => {
    expect(extractDisplayText('')).toBe('(empty)')
  })

  it('returns (empty) for whitespace-only content', () => {
    expect(extractDisplayText('   ')).toBe('(empty)')
  })

  it('returns (empty) for alarm-symbol-only content', () => {
    expect(extractDisplayText('⏰')).toBe('(empty)')
  })

  it('removes alarm directive from display text', () => {
    expect(extractDisplayText('Task [alarm("abc123")]')).toBe('Task')
  })

  it('removes multiple alarm directives', () => {
    expect(extractDisplayText('[alarm("a1")] Task [alarm("a2")]')).toBe('Task')
  })

  it('returns (empty) for alarm-directive-only content', () => {
    expect(extractDisplayText('[alarm("abc")]')).toBe('(empty)')
  })

  it('removes both alarm symbol and alarm directive', () => {
    expect(extractDisplayText('⏰ Task [alarm("id1")]')).toBe('Task')
  })

  it('removes recurring alarm directive from display text', () => {
    expect(extractDisplayText('Task [recurringAlarm("rec123")]')).toBe('Task')
  })

  it('removes multiple recurring alarm directives', () => {
    expect(extractDisplayText('[recurringAlarm("r1")] Task [recurringAlarm("r2")]')).toBe('Task')
  })

  it('returns (empty) for recurringAlarm-directive-only content', () => {
    expect(extractDisplayText('[recurringAlarm("abc")]')).toBe('(empty)')
  })

  it('removes mixed alarm and recurringAlarm directives', () => {
    expect(extractDisplayText('[alarm("a1")] Task [recurringAlarm("r1")]')).toBe('Task')
  })
})
